File: lib/src/main/java/com/jintin/mixadapter/MixAdapter.java
package com.jintin.mixadapter;

import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.RecyclerView.Adapter;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

public class MixAdapter<T extends RecyclerView.ViewHolder> extends Adapter<T> {
    private static final int VIEW_TYPE_OFFSET = 10000;
    private final List<Adapter<? extends T>> adapters;

    public MixAdapter() {
        this.adapters = new ArrayList<>();
    }

    public MixAdapter(final List<Adapter<? extends T>> adapters) {
        this.adapters = adapters;
    }

    public void addAdapter(Adapter<? extends T> adapter) {
        adapters.add(adapter);
        notifyDataSetChanged();
    }

    @Override
    public void onAttachedToRecyclerView(RecyclerView recyclerView) {
        for (Adapter<? extends T> adapter : adapters) {
            adapter.onAttachedToRecyclerView(recyclerView);
        }
    }

    @Override
    public void onDetachedFromRecyclerView(RecyclerView recyclerView) {
        for (Adapter<? extends T> adapter : adapters) {
            adapter.onDetachedFromRecyclerView(recyclerView);
        }
    }

    @Override
    public int getItemViewType(int position) {
        int offset = 0;
        for (int i = 0; i < adapters.size(); i++) {
            Adapter<? extends T> adapter = adapters.get(i);
            if (position < offset + adapter.getItemCount()) {
                return VIEW_TYPE_OFFSET * i + adapter.getItemViewType(position - offset);
            } else {
                offset += adapter.getItemCount();
            }
        }
        throw new RuntimeException("getItemViewType size exceed");
    }

    @Override
    public T onCreateViewHolder(ViewGroup parent, int viewType) {
        int index = viewType / VIEW_TYPE_OFFSET;
        int position = viewType % VIEW_TYPE_OFFSET;
        Adapter<? extends T> adapter = adapters.get(index);
        return adapter.onCreateViewHolder(parent, position);
    }

    @Override
    public void onBindViewHolder(T holder, int position) {
        int offset = 0;
        for (int i = 0; i < adapters.size(); i++) {
            Adapter<T> adapter = (Adapter<T>) adapters.get(i);
            if (position < offset + adapter.getItemCount()) {
                adapter.onBindViewHolder(holder, position - offset);
                return;
            } else {
                offset += adapter.getItemCount();
            }
        }
    }

    @Override
    public int getItemCount() {
        int size = 0;
        for (Adapter<? extends T> adapter : adapters) {
            size += adapter.getItemCount();
        }
        return size;
    }
}

File: lib/src/main/java/com/jintin/mixadapter/MixAdapter.kt
package com.jintin.mixadapter

import android.support.v7.widget.RecyclerView
import android.support.v7.widget.RecyclerView.NO_ID
import android.view.ViewGroup
import java.util.*

/**
 * Compose your RecyclerView.Adapter into MutableList and dynamic dispatch the related function to each correspond RecyclerView.Adapter
 */
class MixAdapter<T : RecyclerView.ViewHolder> : RecyclerView.Adapter<T> {
    private val adapters: MutableList<RecyclerView.Adapter<out T>>

    constructor() {
        this.adapters = ArrayList()
    }

    constructor(adapters: MutableList<RecyclerView.Adapter<out T>>) {
        this.adapters = adapters
    }

    override fun onAttachedToRecyclerView(recyclerView: RecyclerView?) {
        adapters.forEach {
            it.onAttachedToRecyclerView(recyclerView)
        }
    }

    override fun onDetachedFromRecyclerView(recyclerView: RecyclerView?) {
        adapters.forEach {
            it.onDetachedFromRecyclerView(recyclerView)
        }
    }

    override fun setHasStableIds(hasStableIds: Boolean) {
        super.setHasStableIds(hasStableIds)
        adapters.forEach {
            it.setHasStableIds(hasStableIds)
        }
    }

    override fun registerAdapterDataObserver(observer: RecyclerView.AdapterDataObserver?) {
        super.registerAdapterDataObserver(observer)
        adapters.forEach {
            it.registerAdapterDataObserver(observer)
        }
    }

    override fun unregisterAdapterDataObserver(observer: RecyclerView.AdapterDataObserver?) {
        super.unregisterAdapterDataObserver(observer)
        adapters.forEach {
            it.unregisterAdapterDataObserver(observer)
        }
    }

    override fun getItemViewType(position: Int): Int {
        var offset = 0
        var index = position
        adapters.forEach {
            if (index < it.itemCount) {
                return offset + it.getItemViewType(index)
            } else {
                offset += VIEW_TYPE_OFFSET
                index -= it.itemCount
            }
        }
        throw IllegalArgumentException("not found view type in adapters")
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): T {
        val index = viewType / VIEW_TYPE_OFFSET
        val position = viewType % VIEW_TYPE_OFFSET
        val adapter = adapters[index]
        return adapter.onCreateViewHolder(parent, position)
    }

    override fun onBindViewHolder(holder: T, position: Int) {
        var offset = 0
        adapters.filterIsInstance<RecyclerView.Adapter<T>>()
                .forEach {
                    if (position - offset < it.itemCount) {
                        it.onBindViewHolder(holder, position - offset)
                        return
                    } else {
                        offset += it.itemCount
                    }
                }
    }

    override fun getItemCount(): Int = adapters.sumBy { it.itemCount }

    override fun getItemId(position: Int): Long {
        var offset = 0
        adapters.forEach {
            if (position - offset < it.itemCount) {
                return it.getItemId(position - offset)
            } else {
                offset += it.itemCount
            }
        }
        return NO_ID
    }

    /**
     * Add adapter into MixAdapter
     */
    fun addAdapter(adapter: RecyclerView.Adapter<out T>) {
        adapters.add(adapter)
        notifyDataSetChanged()
    }

    /**
     * Get start position of given adapter in MixAdapter
     */
    fun getAdapterOffset(target: RecyclerView.Adapter<*>): Int {
        var offset = 0
        for (adapter in adapters) {
            if (adapter == target) {
                return offset
            }
            offset += adapter.itemCount
        }
        return offset
    }

    companion object {
        private const val VIEW_TYPE_OFFSET = 10000
    }
}
